File: scripts/update_recipes_from_csv.py
#!/usr/bin/env python3
"""
Update recipes.json with ingredient data from CSV.
This script reads the CSV and updates matching recipes in recipes.json
with the correct ingredient data from the CSV.
"""

import csv
import json
import re
from pathlib import Path

def normalize_name(name: str) -> str:
    """Convert a display name to an ID (snake_case)."""
    # Handle special cases
    name = name.strip()
    if not name:
        return ""

    # Handle + suffix for plus quality items
    if name.endswith(' +'):
        name = name[:-2] + '_plus'
    elif name.endswith('+'):
        name = name[:-1] + '_plus'

    # Convert to lowercase and replace spaces/special chars with underscores
    normalized = re.sub(r'[^a-z0-9]+', '_', name.lower()).strip('_')
    return normalized

def parse_ingredient_cell(cell: str) -> list:
    """
    Parse an ingredient cell which may contain:
    - Single ingredient: "Tomato"
    - Multiple choices: "Cheese|Herb Cheese|Cheese +"
    - Empty cell: ""

    Returns a list of normalized ingredient IDs.
    """
    if not cell or not cell.strip():
        return []

    # Split by | for choices
    parts = [p.strip() for p in cell.split('|')]
    return [normalize_name(p) for p in parts if p.strip()]

def build_ingredient_object(ingredients: list) -> dict:
    """
    Build an ingredient object for the recipe.
    If single ingredient, returns {"id": "xxx", "required": true}
    If multiple choices, returns {"id": "choice", "required": true, "oneOf": [...]}
    """
    if len(ingredients) == 0:
        return None
    elif len(ingredients) == 1:
        return {"id": ingredients[0], "required": True}
    else:
        return {"id": "choice", "required": True, "oneOf": ingredients}

def parse_csv_recipe(row: dict) -> dict:
    """Parse a CSV row into recipe ingredient data."""
    recipe_name = row.get('Recipe', '').strip()
    recipe_type = row.get('Type', '').strip()

    if not recipe_name:
        return None

    # Parse required ingredients (Ingredient1-4)
    required_ingredients = []
    for i in range(1, 5):
        col = f'Ingredient{i}'
        if col in row:
            parsed = parse_ingredient_cell(row[col])
            if parsed:
                ing_obj = build_ingredient_object(parsed)
                if ing_obj:
                    required_ingredients.append(ing_obj)

    # Parse additions (IngredientA)
    additions = []
    if 'IngredientA' in row:
        additions = parse_ingredient_cell(row['IngredientA'])

    # Map CSV type to our category format
    category_map = {
        'Salads': 'salad',
        'Soups': 'soup',
        'Sides': 'side',
        'Mains': 'main_dish',
        'Main Dishes': 'main_dish',
        'Desserts': 'dessert',
        'Other': 'other',
        'Other Recipes': 'other'
    }

    # Parse additional fields
    base_value = row.get('Sales Price', '0').replace(',', '')
    try:
        base_value = int(base_value)
    except ValueError:
        base_value = 0

    effect = None
    effect_level = None
    effect_raw = row.get('Effect', '').strip()
    if effect_raw and effect_raw != 'None':
        # Parse effect like "Stamina Saver Lv. 3"
        effect_match = re.match(r'(.+?)\s*Lv\.?\s*(\d+)', effect_raw)
        if effect_match:
            effect_name = effect_match.group(1).strip()
            effect_level = int(effect_match.group(2))
            # Convert effect name to ID
            effect_map = {
                'Stamina Saver': 'stamina_saver',
                'Watering Rare Crop %': 'watering_rare_crop',
                'Rod Durability Up': 'rod_durability_up',
                'Run Speed Up': 'run_speed_up',
                'Max Stamina Up': 'max_stamina_up',
                'Chat Friendship Boost': 'chat_friendship_boost',
                'Petting Happiness Up': 'petting_happiness_up',
                'Petting Rare Item %': 'petting_rare_item',
                'Petting Rare Item': 'petting_rare_item',
                'Pet Training Up': 'pet_training_up',
                'Happy Energy Gain': 'happy_energy_gain',
            }
            effect = effect_map.get(effect_name, normalize_name(effect_name))

    result = {
        'name': recipe_name,
        'id': normalize_name(recipe_name),
        'category': category_map.get(recipe_type, 'other'),
        'ingredients': required_ingredients,
        'additions': additions,
        'baseValue': base_value,
    }

    if effect:
        result['effect'] = effect
    if effect_level:
        result['effectLevel'] = effect_level

    return result

def load_csv(csv_path: str) -> list:
    """Load and parse the CSV file."""
    recipes = []
    with open(csv_path, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            parsed = parse_csv_recipe(row)
            if parsed:
                recipes.append(parsed)
    return recipes

def update_recipes_json(json_path: str, csv_recipes: list) -> tuple:
    """
    Update recipes.json with data from CSV.
    Updates ingredients, additions, baseValue, effect, effectLevel.
    Also adds new recipes that don't exist yet.
    Preserves other fields like unlockMethod, icon, etc.

    Returns (updated_count, added_count, not_found_recipes)
    """
    with open(json_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    # Create lookup by normalized name
    csv_lookup = {r['id']: r for r in csv_recipes}
    # Also create lookup by exact name for fallback matching
    csv_name_lookup = {r['name'].lower(): r for r in csv_recipes}

    # Track which CSV recipes we've matched
    matched_csv_ids = set()

    updated_count = 0
    not_found = []

    for recipe in data['recipes']:
        recipe_id = recipe['id']
        recipe_name = recipe.get('name', '').lower()

        csv_recipe = None

        # Try to find match by ID first
        if recipe_id in csv_lookup:
            csv_recipe = csv_lookup[recipe_id]
            matched_csv_ids.add(recipe_id)
        # Fallback: try by name
        elif recipe_name in csv_name_lookup:
            csv_recipe = csv_name_lookup[recipe_name]
            matched_csv_ids.add(csv_recipe['id'])

        if csv_recipe:
            # Update ingredients
            recipe['ingredients'] = csv_recipe['ingredients']

            # Update additions (or remove if empty)
            if csv_recipe['additions']:
                recipe['additions'] = csv_recipe['additions']
            elif 'additions' in recipe:
                del recipe['additions']

            # Update baseValue
            if csv_recipe['baseValue']:
                recipe['baseValue'] = csv_recipe['baseValue']

            # Update effect/effectLevel
            if csv_recipe.get('effect'):
                recipe['effect'] = csv_recipe['effect']
            if csv_recipe.get('effectLevel'):
                recipe['effectLevel'] = csv_recipe['effectLevel']

            updated_count += 1
            print(f"Updated: {recipe['name']}")
        else:
            not_found.append(recipe['name'])

    # Add new recipes from CSV that don't exist in JSON
    added_count = 0
    for csv_recipe in csv_recipes:
        if csv_recipe['id'] not in matched_csv_ids:
            # Create new recipe entry
            new_recipe = {
                'id': csv_recipe['id'],
                'name': csv_recipe['name'],
                'category': csv_recipe['category'],
                'utensil': 'none',  # Default, will need manual update
                'ingredients': csv_recipe['ingredients'],
                'baseValue': csv_recipe['baseValue'],
                'unlockMethod': 'Unknown',
                'icon': f"{csv_recipe['id']}.png"
            }

            if csv_recipe['additions']:
                new_recipe['additions'] = csv_recipe['additions']

            if csv_recipe.get('effect'):
                new_recipe['effect'] = csv_recipe['effect']
            if csv_recipe.get('effectLevel'):
                new_recipe['effectLevel'] = csv_recipe['effectLevel']

            data['recipes'].append(new_recipe)
            added_count += 1
            print(f"Added new: {csv_recipe['name']}")

    # Sort recipes by category then name
    category_order = ['salad', 'soup', 'side', 'main_dish', 'dessert', 'other']
    data['recipes'].sort(key=lambda r: (
        category_order.index(r['category']) if r['category'] in category_order else 99,
        r['name']
    ))

    # Save updated JSON
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2)

    return updated_count, added_count, not_found

def main():
    script_dir = Path(__file__).parent
    project_dir = script_dir.parent

    csv_path = project_dir / "SOS_ Grand Bazaar Recipes - Sheet1.csv"
    recipes_json_path = project_dir / "public" / "data" / "recipes.json"
    data_json_path = project_dir / "public" / "data" / "data.json"

    print(f"Loading CSV from: {csv_path}")
    csv_recipes = load_csv(csv_path)
    print(f"Loaded {len(csv_recipes)} recipes from CSV")

    # Print some examples for verification
    print("\nSample parsed CSV recipes:")
    for r in csv_recipes[:3]:
        print(f"  {r['name']}: {len(r['ingredients'])} ingredients, {len(r['additions'])} additions")
        if r.get('effect'):
            print(f"    Effect: {r['effect']} Lv. {r.get('effectLevel', '?')}")

    # Update data.json (this is the main data file now)
    if data_json_path.exists():
        print(f"\nUpdating data.json at: {data_json_path}")
        updated, added, not_found = update_recipes_json(str(data_json_path), csv_recipes)
        print(f"\nUpdated {updated} recipes, added {added} new recipes in data.json")

        if not_found:
            print(f"\nRecipes in JSON not found in CSV ({len(not_found)}):")
            for name in sorted(not_found):
                print(f"  - {name}")

    # Also update recipes.json if it exists
    if recipes_json_path.exists():
        print(f"\nUpdating recipes.json at: {recipes_json_path}")
        updated2, added2, _ = update_recipes_json(str(recipes_json_path), csv_recipes)
        print(f"Updated {updated2} recipes, added {added2} new recipes")

if __name__ == '__main__':
    main()

File: scripts/update-recipes-from-csv.mjs
#!/usr/bin/env node
/**
 * Update recipes.json from Google Sheets CSV export.
 * Preserves utensil, unlockMethod, icon from existing data.
 * Updates ingredients, price, effect, additions, and adds new fields.
 */

import { readFileSync, writeFileSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const DATA_DIR = resolve(__dirname, '../public/data');

// Load existing data
const existingRecipes = JSON.parse(readFileSync(resolve(DATA_DIR, 'recipes.json'), 'utf8')).recipes;
const ingredients = JSON.parse(readFileSync(resolve(DATA_DIR, 'ingredients.json'), 'utf8')).ingredients;
const effects = JSON.parse(readFileSync(resolve(DATA_DIR, 'effects.json'), 'utf8')).effects;
const characters = JSON.parse(readFileSync(resolve(DATA_DIR, 'characters.json'), 'utf8')).characters;

// Build lookup maps
const ingredientNameToId = new Map();
ingredients.forEach(ing => {
  ingredientNameToId.set(ing.name.toLowerCase().trim(), ing.id);
});

// Also map "+" suffix variants: "Milk +" -> "milk_plus", "Egg +" -> "egg_plus"
ingredients.forEach(ing => {
  if (ing.id.endsWith('_plus')) {
    // Create alternate name mapping: "milk+" and "milk +"
    const baseName = ing.name.replace(/\+$/, '').trim();
    ingredientNameToId.set((baseName + '+').toLowerCase(), ing.id);
    ingredientNameToId.set((baseName + ' +').toLowerCase(), ing.id);
  }
});

const effectNameToId = new Map();
effects.forEach(eff => {
  effectNameToId.set(eff.name.toLowerCase().trim(), eff.id);
});

const characterNameToId = new Map();
characters.forEach(ch => {
  characterNameToId.set(ch.name.toLowerCase().trim(), ch.id);
});

// Normalize accented characters for matching
function normalizeAccents(str) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

// Build existing recipe map by name (for matching)
const existingByName = new Map();
const existingById = new Map();
existingRecipes.forEach(r => {
  existingByName.set(r.name.toLowerCase().trim(), r);
  existingByName.set(normalizeAccents(r.name.toLowerCase().trim()), r);
  existingById.set(r.id, r);
});

// Category mapping from CSV to our format
const categoryMap = {
  'salads': 'salad',
  'soups': 'soup',
  'sides': 'side',
  'mains': 'main_dish',
  'desserts': 'dessert',
  'other': 'other',
};

// Convert a display name to an ingredient ID
function nameToIngredientId(name) {
  name = name.trim();
  if (!name) return null;

  // Fix known typos from spreadsheet
  const typoFixes = {
    'shiitake mushrrom': 'shiitake mushroom',
    'silkit egg': 'silkie egg',
    'banana tea tin': 'banana tea tin',
    'mango juiice': 'mango juice',
    'grilled fish': 'grilled fish',
  };

  const lower = name.toLowerCase().trim();
  if (typoFixes[lower]) {
    name = typoFixes[lower];
  }

  const normalized = name.toLowerCase().trim();

  // Direct lookup
  if (ingredientNameToId.has(normalized)) {
    return ingredientNameToId.get(normalized);
  }

  // Try with "+" normalization: "Milk +" -> look for "milk+"
  if (normalized.endsWith('+')) {
    const withoutSpace = normalized.replace(/\s*\+$/, '+');
    if (ingredientNameToId.has(withoutSpace)) {
      return ingredientNameToId.get(withoutSpace);
    }
    const withSpace = normalized.replace(/\s*\+$/, ' +');
    if (ingredientNameToId.has(withSpace)) {
      return ingredientNameToId.get(withSpace);
    }
  }

  // Fallback: convert to snake_case ID
  let id = normalized
    .replace(/\+/g, '_plus')
    .replace(/['']/g, '')
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_|_$/g, '');

  // Handle "+" at end: "Milk +" -> "milk_plus"
  if (name.trim().endsWith('+') && !id.endsWith('_plus')) {
    id = id.replace(/_$/, '') + '_plus';
  }

  return id;
}

// Convert recipe name to ID
function nameToRecipeId(name) {
  return name.toLowerCase().trim()
    .replace(/[''&]/g, '')
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_|_$/g, '');
}

// Parse effect string like "Chat Friendship Boost Lv. 4" -> { id, level }
function parseEffect(effectStr) {
  if (!effectStr || effectStr.toLowerCase() === 'none') {
    return { id: null, level: null };
  }

  const match = effectStr.match(/^(.+?)\s+Lv\.\s*(\d+)$/);
  if (match) {
    const effectName = match[1].trim().toLowerCase();
    const level = parseInt(match[2]);

    // Find effect by name
    for (const [name, id] of effectNameToId) {
      if (name === effectName || effectName.includes(name) || name.includes(effectName)) {
        return { id, level };
      }
    }

    // Try matching by converting to ID format
    const effectId = effectName.replace(/[^a-z0-9]+/g, '_').replace(/_%$/, '');
    const found = effects.find(e => e.id === effectId);
    if (found) return { id: found.id, level };

    // Fuzzy match
    const possibleId = effectName.replace(/ %/g, '').replace(/[^a-z0-9]+/g, '_');
    const found2 = effects.find(e => e.id === possibleId);
    if (found2) return { id: found2.id, level };

    console.warn(`  WARNING: Unknown effect: "${effectStr}" -> tried "${possibleId}"`);
    return { id: possibleId, level };
  }

  // No level (e.g., "Double Forage", "Jump Distance Up")
  const effectName = effectStr.trim().toLowerCase();
  for (const [name, id] of effectNameToId) {
    if (name === effectName) {
      return { id, level: 1 };
    }
  }

  const effectId = effectName.replace(/[^a-z0-9]+/g, '_');
  return { id: effectId, level: 1 };
}

// Parse a pipe-separated ingredient cell into ingredient entries
// "Cheese|Herb Cheese|Cheese +" -> oneOf group
// "Potato" -> single ingredient
function parseIngredientCell(cell) {
  if (!cell || !cell.trim()) return null;

  // Fix separator typos: } and \ used instead of |
  cell = cell.replace(/\}/g, '|').replace(/\\/g, '|');

  const parts = cell.split('|').map(p => p.trim()).filter(Boolean);

  if (parts.length === 0) return null;

  if (parts.length === 1) {
    const id = nameToIngredientId(parts[0]);
    return { id, required: true };
  }

  // Multiple options -> oneOf
  const ids = parts.map(p => nameToIngredientId(p));
  return { id: 'choice', required: true, oneOf: ids };
}

// Parse a pipe-separated additions cell
function parseAdditions(cell) {
  if (!cell || !cell.trim()) return [];

  // Fix separator typos
  cell = cell.replace(/\}/g, '|').replace(/\\/g, '|');

  return cell.split('|').map(p => p.trim()).filter(Boolean).map(nameToIngredientId);
}

// Parse character names from gift columns
function parseCharacterNames(cell) {
  if (!cell || !cell.trim()) return [];
  return cell.split('|').map(n => n.trim()).filter(Boolean);
}

// Parse CSV (handle quoted fields)
function parseCSV(csvText) {
  const rows = [];
  const lines = csvText.split('\n');
  let currentRow = [];
  let inQuotes = false;
  let currentField = '';

  for (const line of lines) {
    for (let i = 0; i < line.length; i++) {
      const ch = line[i];
      if (inQuotes) {
        if (ch === '"') {
          if (i + 1 < line.length && line[i + 1] === '"') {
            currentField += '"';
            i++;
          } else {
            inQuotes = false;
          }
        } else {
          currentField += ch;
        }
      } else {
        if (ch === '"') {
          inQuotes = true;
        } else if (ch === ',') {
          currentRow.push(currentField);
          currentField = '';
        } else {
          currentField += ch;
        }
      }
    }

    if (inQuotes) {
      currentField += '\n';
    } else {
      currentRow.push(currentField);
      currentField = '';
      if (currentRow.some(f => f.trim())) {
        rows.push(currentRow);
      }
      currentRow = [];
    }
  }

  if (currentRow.length > 0) {
    currentRow.push(currentField);
    if (currentRow.some(f => f.trim())) {
      rows.push(currentRow);
    }
  }

  return rows;
}

// Main processing
const csvText = readFileSync(resolve(__dirname, 'recipes-spreadsheet.csv'), 'utf8');
const rows = parseCSV(csvText);

// Header row
const headers = rows[0];
console.log('CSV Headers:', headers);
console.log(`Processing ${rows.length - 1} recipe rows...`);

const updatedRecipes = new Map(); // id -> recipe
const giftUpdates = []; // Track gift data changes
const unmatchedIngredients = new Set();

// Process each CSV row
for (let i = 1; i < rows.length; i++) {
  const row = rows[i];
  if (row.length < 11) continue;

  const [type, recipeName, salesPrice, effectStr, staminaRecovery, holidayGift, bazaarTrend, favoriteGift, lovedGift, likedGift, ing1, ing2, ing3, ing4, ingA] = row;

  if (!recipeName || !recipeName.trim()) continue;

  const recipeId = nameToRecipeId(recipeName);
  const existing = existingByName.get(recipeName.toLowerCase().trim()) || existingByName.get(normalizeAccents(recipeName.toLowerCase().trim())) || existingById.get(recipeId);

  // Parse category
  const category = categoryMap[type.trim().toLowerCase()] || 'other';

  // Parse effect
  const { id: effectId, level: effectLevel } = parseEffect(effectStr?.trim());

  // Parse ingredients (1-4)
  const ingredientEntries = [];
  for (const cell of [ing1, ing2, ing3, ing4]) {
    const parsed = parseIngredientCell(cell);
    if (parsed) ingredientEntries.push(parsed);
  }

  // Parse additions
  const additions = parseAdditions(ingA);

  // Build recipe object
  const recipe = {
    id: existing?.id || recipeId,
    name: recipeName.trim(),
    category,
    utensil: existing?.utensil || 'none',
    ingredients: ingredientEntries,
    ...(additions.length > 0 ? { additions } : {}),
    baseValue: parseInt(salesPrice) || 0,
    effect: effectId || null,
    effectLevel: effectLevel || null,
    unlockMethod: existing?.unlockMethod || '',
    icon: existing?.icon || `${recipeId}.png`,
    ...(staminaRecovery?.trim() ? { staminaRecovery: staminaRecovery.trim() } : {}),
    ...(holidayGift?.trim() ? { holidayGift: holidayGift.trim() } : {}),
    ...(bazaarTrend?.trim() ? { bazaarTrend: bazaarTrend.trim() } : {}),
  };

  // Clean null effect fields
  if (!recipe.effect) {
    delete recipe.effect;
    delete recipe.effectLevel;
  }

  updatedRecipes.set(recipe.id, recipe);

  // Track gift data
  const favChars = parseCharacterNames(favoriteGift);
  const lovedChars = parseCharacterNames(lovedGift);
  const likedChars = parseCharacterNames(likedGift);

  if (favChars.length || lovedChars.length || likedChars.length) {
    giftUpdates.push({
      recipeId: recipe.id,
      recipeName: recipe.name,
      favorite: favChars,
      loved: lovedChars,
      liked: likedChars,
    });
  }
}

// Merge: keep existing recipes not in CSV, update ones that are
const finalRecipes = [];
const csvRecipeIds = new Set(updatedRecipes.keys());
let updatedCount = 0;
let newCount = 0;
let preservedCount = 0;

// First add all CSV recipes (these are the updated ones)
for (const [id, recipe] of updatedRecipes) {
  if (existingById.has(id) || existingByName.has(recipe.name.toLowerCase().trim())) {
    updatedCount++;
  } else {
    newCount++;
    console.log(`  NEW: ${recipe.name} (${id})`);
  }
  finalRecipes.push(recipe);
}

// Then add existing recipes NOT in the CSV (preserve them)
for (const existing of existingRecipes) {
  if (!csvRecipeIds.has(existing.id)) {
    // Check by name too
    const inCSV = Array.from(updatedRecipes.values()).some(
      r => normalizeAccents(r.name.toLowerCase().trim()) === normalizeAccents(existing.name.toLowerCase().trim())
    );
    if (!inCSV) {
      finalRecipes.push(existing);
      preservedCount++;
    }
  }
}

// Sort by category then name
const categoryOrder = ['salad', 'soup', 'side', 'main_dish', 'dessert', 'other'];
finalRecipes.sort((a, b) => {
  const catDiff = categoryOrder.indexOf(a.category) - categoryOrder.indexOf(b.category);
  if (catDiff !== 0) return catDiff;
  return a.name.localeCompare(b.name);
});

// Write updated recipes.json
writeFileSync(
  resolve(DATA_DIR, 'recipes.json'),
  JSON.stringify({ recipes: finalRecipes }, null, 2) + '\n'
);

console.log('\n=== Summary ===');
console.log(`Updated: ${updatedCount} recipes`);
console.log(`New: ${newCount} recipes`);
console.log(`Preserved (not in CSV): ${preservedCount} recipes`);
console.log(`Total: ${finalRecipes.length} recipes`);

// Update character gift data from CSV
if (giftUpdates.length > 0) {
  console.log(`\n=== Gift Updates (${giftUpdates.length} recipes with gift data) ===`);

  // Build a map of character name -> character object for updating
  const charByName = new Map();
  characters.forEach(ch => charByName.set(ch.name.toLowerCase(), ch));

  for (const gift of giftUpdates) {
    const parts = [];
    if (gift.favorite.length) parts.push(`Fav: ${gift.favorite.join(', ')}`);
    if (gift.loved.length) parts.push(`Loved: ${gift.loved.join(', ')}`);
    if (gift.liked.length) parts.push(`Liked: ${gift.liked.join(', ')}`);
    console.log(`  ${gift.recipeName}: ${parts.join(' | ')}`);

    // Update character records
    for (const charName of gift.favorite) {
      const ch = charByName.get(charName.toLowerCase());
      if (ch) {
        if (!ch.favoriteGifts) ch.favoriteGifts = [];
        if (!ch.favoriteGifts.includes(gift.recipeId)) {
          ch.favoriteGifts.push(gift.recipeId);
        }
      } else {
        console.warn(`  WARNING: Unknown character "${charName}" for favorite gift`);
      }
    }

    for (const charName of gift.loved) {
      const ch = charByName.get(charName.toLowerCase());
      if (ch) {
        if (!ch.lovedGifts) ch.lovedGifts = [];
        if (!ch.lovedGifts.includes(gift.recipeId)) {
          ch.lovedGifts.push(gift.recipeId);
        }
      } else {
        console.warn(`  WARNING: Unknown character "${charName}" for loved gift`);
      }
    }

    for (const charName of gift.liked) {
      const ch = charByName.get(charName.toLowerCase());
      if (ch) {
        if (!ch.likedGifts) ch.likedGifts = [];
        if (!ch.likedGifts.includes(gift.recipeId)) {
          ch.likedGifts.push(gift.recipeId);
        }
      } else {
        console.warn(`  WARNING: Unknown character "${charName}" for liked gift`);
      }
    }
  }

  // Write updated characters.json
  writeFileSync(
    resolve(DATA_DIR, 'characters.json'),
    JSON.stringify({ characters }, null, 2) + '\n'
  );
  console.log('Updated characters.json with gift data');
}

// Validate ingredient IDs exist
const allIngIds = new Set(ingredients.map(i => i.id));
let missingIngCount = 0;
for (const recipe of finalRecipes) {
  for (const ing of recipe.ingredients) {
    if (ing.oneOf) {
      for (const id of ing.oneOf) {
        if (!allIngIds.has(id)) {
          unmatchedIngredients.add(id);
          missingIngCount++;
        }
      }
    } else if (ing.id && ing.id !== 'choice') {
      if (!allIngIds.has(ing.id)) {
        unmatchedIngredients.add(ing.id);
        missingIngCount++;
      }
    }
  }
  if (recipe.additions) {
    for (const id of recipe.additions) {
      if (!allIngIds.has(id)) {
        unmatchedIngredients.add(id);
      }
    }
  }
}

if (unmatchedIngredients.size > 0) {
  console.log(`\n=== Missing Ingredient IDs (${unmatchedIngredients.size}) ===`);
  for (const id of [...unmatchedIngredients].sort()) {
    console.log(`  ${id}`);
  }
}

console.log('\nDone!');
